### PR TITLE
gnrc_tcp: rewrite recv buffer allocation/returning

### PR DESCRIPTION
**Pull request by brummer-simon**
Hi fellow RIOTers,

This PR moves the allocation and release of a connections receive buffer from the internal FSM to the gnrc_tcp functions related to open and close.

Just like #14222, this is a preparation for an API change in gnrc_tcp (implements a listen/accept scheme) to simplify sock_tcp integration.  I have tested this PR on native.

This PR does not introduce new features, testing by executing the test scripts in "tests/gnrc_tcp" should be sufficient.

### CODE REVIEW

**Comment by brummer-simon** [comment]
To clarify the Issue with this PR. Receive buffer allocation is only required on opening a connection. This occurs only by calling gnrc_tcp_open_active or gnrc_tcp_open_passive. Having the receive buffer allocation outside of the FSM is basically required to implement sock_tcp_listen / sock_tcp_stop_listen in a sane way later on.

However releasing the receive buffer is required at every transition to a closed connection state. This can happen basically everywhere within gnrc_tcp (and it is easy to craft packets that trigger this), thats why it was part of the transition into the closed state. 

To solve this I am proposing is the following:
- Receive buffer allocation within "gnrc_tcp_open_active" and "gnrc_tcp_open_passive".
- Receive buffer returning within the FSM while transitioning to a closed state.

This would solve the Issues I have and provide safe memory handling with the drawback of splitting memory handling over two compilation units (might be confusing for the readers).

@miri64 - As a reviewer, are you okay with that approach?

**Comment by brummer-simon** [comment]
I solved the situation by moving the memory allocation from the entry of state LISTEN/SYN_SEND to the connection opening function. This simplifies this PR significantly and solves the issue I had with my previous approach.

File: sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c [hunk 1]
```diff
@@ -148,11 +148,6 @@ static int _transition_to(gnrc_tcp_tcb_t *tcb, fsm_state_t state)
 #endif
             tcb->peer_port = PORT_UNSPEC;
 
-            /* Allocate receive buffer */
-            if (_rcvbuf_get_buffer(tcb) == -ENOMEM) {
```
Comment (miri64) [inline sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c:152]: Maybe now that only `FSM_STATE_SYN_SENT` may fail, add this fact to the documentation of `_transition_to()`:

```
@return   -EADDRINUSE, if @p state == `FSM_STATE_SYN_SENT` and local addres is already in use
```

**Comment by miri64** [comment]
Need to retest. Code-design still looks sane.

**Comment by miri64** [comment]
Can you give commit and PR maybe a more human readable title, please!

**Comment by brummer-simon** [comment]
Better?

**Comment by miri64** [comment]
Yepp :-).

**Comment by miri64** [comment]
(ACK upholds)

**Comment by miri64** [comment]
Thanks for the contribution!

**Pull request merged**